File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/details/router/PostDetailsActivity.java
package com.david_kapas.android.techtest.presentation.details.router;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;

import com.david_kapas.android.techtest.R;
import com.david_kapas.android.techtest.databinding.ActivityPostDetailsBinding;
import com.david_kapas.android.techtest.presentation.details.viewmodel.PostDetailsViewModel;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

/**
 * Main screen for showing post details.
 * Created by David Kapas on 2018.03.17.
 */

public class PostDetailsActivity extends AppCompatActivity implements PostDetailsRouter {

    public static final String POST_ID_PARAM = "postIdParam";

    @Inject
    PostDetailsViewModel postDetailsViewModel;

    private int postId;
    private boolean injected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        //inject();
        ActivityPostDetailsBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_post_details);
        binding.setViewModel(postDetailsViewModel);
        if (getIntent() != null) {
            postId = getIntent().getIntExtra(POST_ID_PARAM, -1);
        }
        if (postId > 0) {
            postDetailsViewModel.showPostDetails(postId);
        }

    }

    /*  private void inject() {
        if (!injected) {
            PostDetailsActivityComponent.Injector.buildComponent(this).inject(this);
            injected = true;
        }
    }*/

    @Override
    public void showGeneralErrorDialog() {
        showNetworkErrorDialog();
    }

    private void showNetworkErrorDialog() {
        AlertDialog alertDialog = new AlertDialog.Builder(PostDetailsActivity.this).create();
        alertDialog.setTitle(R.string.network_error_title);
        alertDialog.setMessage(getString(R.string.network_error_message));
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "OK",
                (dialog, which) -> dialog.dismiss());
        alertDialog.show();
    }
}

File: TechTest/app/src/main/java/com/david_kapas/android/techtest/presentation/details/router/PostDetailsActivity.kt
package com.david_kapas.android.techtest.presentation.details.router

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import com.david_kapas.android.techtest.R
import com.david_kapas.android.techtest.databinding.ActivityPostDetailsBinding
import com.david_kapas.android.techtest.di.modules.PostDetailsActivityModule
import com.david_kapas.android.techtest.presentation.details.viewmodel.PostDetailsViewModel
import dagger.android.AndroidInjection
import javax.inject.Inject

/**
 * Main screen for showing post details.
 * Created by David Kapas on 3/26/2019.
 */
class PostDetailsActivity : AppCompatActivity() {


    companion object STATIC {
        public val POST_ID_PARAM = "postIdParam"
    }

    @Inject
    lateinit var postDetailsViewModelFactory: PostDetailsActivityModule.PostDetailsViewModelFactory;
    lateinit var postDetailsViewModel: PostDetailsViewModel;

    private var postId: Int = 0
    val alertDialog: AlertDialog by lazy { createAlertDialog() }

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState)
        val binding = DataBindingUtil.setContentView<ActivityPostDetailsBinding>(this, R.layout.activity_post_details)
        postDetailsViewModel = ViewModelProviders.of(this, postDetailsViewModelFactory).get(PostDetailsViewModel::class.java);
        binding.viewModel = postDetailsViewModel;
        postDetailsViewModel.showErrorMessage.observe(this, Observer { showGeneralErrorDialog() })
        if (intent != null) {
            postId = intent.getIntExtra(POST_ID_PARAM, -1)
        }
        if (postId > 0) {
            postDetailsViewModel.showPostDetails(postId)
        }

    }


    fun showGeneralErrorDialog() {
        showNetworkErrorDialog();
    }

    private fun createAlertDialog(): AlertDialog {
        val alertDialog = AlertDialog.Builder(this@PostDetailsActivity).create()
        alertDialog.setTitle(R.string.network_error_title)
        alertDialog.setMessage(getString(R.string.network_error_message))
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "OK"
        ) { dialog, which -> dialog.dismiss() }
        return alertDialog;
    }

    private fun showNetworkErrorDialog() {
        alertDialog.show()
    }
}